README.md
# flight-request

[![Build Status](https://secure.travis-ci.org/giuseppegurgone/flight-request.png)](http://travis-ci.org/giuseppegurgone/flight-request)

A [Flight](https://github.com/flightjs/flight) mixin for making XHR requests.

## Installation

```bash
bower install --save flight-request
```

## Example
```bash
define(function (require) {
	var defineComponent = require('flight/lib/component');
  	var withRequest = require('flight-request/lib/with_request');

	return defineComponent(dataComponent, withRequest);

	function dataComponent() {
		this.after('initialize', function () {
			this.get({
				url: 'http://b.ar/gimme-beer',
				success: function () {
					console.log('Skål!');
				}
			});
		});
	}
});
```

## Methods and Events
### Methods
flight-request exposes the following methods:

* `this.get`, `this.post`, `this.put`, `this.destroy` (you can also use `this.delete` if you are in a ES5 env) - return a jqXHR object [*].
* `this.abort` takes a jqXHR object as parameter and aborts the given request.
* `this.abortAllRequests` - to abort all the open xhr requests.

[*] by default the requests can be aborted either with this.abort or this.abortAllRequest. If you want to prevent this to happen you can set `noAbort` to true and pass it along with the ajax settings

#### callbacks context
From v1.0.0 the default callbacks context is the component one.
However you can change it by setting `context` in the ajax settings or using bind.

```javascript
// by default the callbacks context is the component one
this.get({
	url: 'http://b.ar/gimme-beer',
	success: function () {
		// this === component
		this.doSomething();
	}
});


// the context can be changed with bind or by setting the context property
this.get({
	url: 'http://b.ar/gimme-beer',
	context: 'foo',
	success: function () {
		// this === 'foo'
	},
	error: function () {
		// this === 'bar'
	}.bind('bar')
});
```

#### tunnelMethod - PUT and DELETE request tunnelling
<blockquote>HTML forms (up to HTML version 4 and XHTML 1) only support GET and POST as HTTP request methods. 
A workaround for this is to tunnel other methods (PUT and DELETE) through POST by using a hidden form field 
which is read by the server and the request dispatched accordingly.</blockquote>

[http://stackoverflow.com/a/166501/931594](http://stackoverflow.com/a/166501/931594)

Method tunnelling is enabled by default for PUT and DELETE requests.
In those cases the request data object contains a `_method` property with the original request type (verb).<br>
From v1.0.0 it is possible to disable the tunnelling by setting `tunnelMethod` to `false` in the ajax settings.<br>
It is also possible to change the tunnelling property name by setting `tunnelMethod` to a string of your choice (one time change)<br>
or via the component settings `this.attr.tunnelMethod`.


### Events
You can specify custom events names (instead of functions) to be triggered on `success` and/or `error`.
Thanks to Angus Croll for the [suggestion](https://github.com/giuseppeg/flight-request/issues/1).

```javascript
this.on('thereYouGo', function (event, responseData, textStatus, jqXHR) {
	console.log('Skål!');
});

this.get({
	url: 'http://b.ar/gimme-beer',
	success: 'thereYouGo'
});
```

All the requests are automatically aborted when the window is unloading its content and resources (window.onunload);
…

## Development

Development of this component requires [Bower](http://bower.io) to be globally
installed:

```bash
npm install -g bower
```

Then install the Node.js and client-side dependencies by running the following
commands in the repo's root directory.

```bash
npm install & bower install
```

To continuously run the tests in Chrome during development, just run:

```bash
npm run watch-test
```

## Contributing to this project

Anyone and everyone is welcome to contribute. Please take a moment to
review the [guidelines for contributing](CONTRIBUTING.md).

* [Bug reports](CONTRIBUTING.md#bugs)
* [Feature requests](CONTRIBUTING.md#features)
* [Pull requests](CONTRIBUTING.md#pull-requests)


CHANGELOG.md
## v1.0.0 (11/2/2014)

### Fixes 
* #5 set default callbacks context to the component one
* #6 use this.attributes 
* #7 possibility to disable the method tunneling for PUT and DELETE requests
* add tests

### Features
* can set the callbacks context in the request settings


## v0.0.2 (3/22/2014)

### Fixes 
* abortAllRequests should loop through the requests array
* HTTP request tunneling workaround breaks setting request type in options/settings object

### Features
* trigger custom events on success/error

test/spec/with_request.spec.js
define(function (require) {
  'use strict';

  var defineComponent = require('flight/lib/component');
  var withRequest = require('lib/with_request');

  describeComponent('lib/with_request', function () {

    var testUrl = '/test';
    var request;
    var testResponses = {
      success: {
        status: 200,
        statusText: 'success',
        responseText: '{"flight":"rocks"}'
      },
      error: {
        status: 400,
        statusText: 'error',
        responseText: '{"status":"400", "statuText": "error", "message":"horror"}'
      }
    };

    var component;    

    function withBaseComponent() { 
      this.testSuccessMethod = jasmine.createSpy('testSuccessMethod');
      this.testErrorMethod = jasmine.createSpy('testErrorMethod');

      this.mockRequest = function (method, settings) {
        this[method](settings);
        return jasmine.Ajax.requests.mostRecent();
      };
    }

    function makeComponent(component) {
      return (new (defineComponent(component, withBaseComponent, withRequest))).initialize(document);
    }

    describe('basic suite', function () {      
      beforeEach(function () {
        jasmine.Ajax.install();
        component = makeComponent(function () {          
          this.after('initialize', function () {
            this.get({
              url: testUrl,
              success: this.testSuccessMethod,
              error: this.testErrorMethod
            });
          });
        });

        request = jasmine.Ajax.requests.mostRecent();
      });

      afterEach(function () {
        jasmine.Ajax.uninstall();
        request = undefined;
        component && component.teardown();
        component = undefined;
      });

      it('should be mixed in', function () {
        [
          'get',
          'post',
          'put',
          'destroy',
          'delete',
          'abortAllRequests'
        ].forEach(function (method) {
          expect(typeof component[method]).toBe("function");
        }, null);
      });

      it('should call the success callback on success', function () {
        request.response(testResponses.success);
        expect(component.testSuccessMethod).toHaveBeenCalled();
      });

      it('should call the error callback on failure', function () {
        request.response(testResponses.error);
        expect(component.testErrorMethod).toHaveBeenCalled();
      });

      it('should abort requests', function () {        
        component.abortAllRequests();
        expect(request.readyState).toBe(0);
        expect(request.statusText).toBe('abort');
        expect(component.testErrorMethod.calls[0].args[0].isAbort).toBe(true);
      });

      it('should set the default callbacks context to component', function () {
        component.mockRequest('get', {
          url: testUrl,
          success: function (data) {
            expect(this).toBe(component);
          }
        }).response(testResponses.success);

        component.mockRequest('get', {
          url: testUrl,
          error: function (data) {
            expect(this).toBe(component);
          }
        }).response(testResponses.error);
      });

      it('should change the callbacks context when provided with the settings or set with bind', function () {
        component.mockRequest('get', {
          url: testUrl,
          context: 'test',
          success: function (data) {
            expect(this).toBe('test');
          }
        }).response(testResponses.success);

        // bind wins
        component.mockRequest('get', {
          url: testUrl,
          context: 'test',
          success: function (data) {
            expect(this).toBe('testBind');
          }.bind('testBind')
        }).response(testResponses.success);

        component.mockRequest('get', {
          url: testUrl,
          success: function (data) {
            expect(this).toBe('test');
          }.bind('test')
        }).response(testResponses.success);
      });

      it('should tunnel PUT and DELETE requests by default', function () {
        request = component.mockRequest('put', {});
        expect(request.method).toBe('POST');

        request = component.mockRequest('delete', {});
        expect(request.method).toBe('POST');
      });

      it('shouldn\'t tunnel PUT and DELETE requests when tunnelMethod is false', function () {
        request = component.mockRequest('put', {
          tunnelMethod: false
        });
        expect(request.method).toBe('PUT');

        request = component.mockRequest('delete', {
          tunnelMethod: false
        });
        expect(request.method).toBe('DELETE');
      });
    });

    describe('events:', function () {
      beforeEach(function () {
        jasmine.Ajax.install();
        component = makeComponent(function () {
          this.attributes({
            successEvent: 'requestSucceed',
            errorEvent: 'requestFailed'
          });

          this.after('initialize', function () {
            this.on(this.attr.successEvent, this.testSuccessMethod);
            this.on(this.attr.errorEvent, this.testErrorMethod);

            this.get({
              url: testUrl,
              success: this.attr.successEvent,
              error: this.attr.errorEvent
            });
          });
        });

        request = jasmine.Ajax.requests.mostRecent();
      });

      afterEach(function () {
        jasmine.Ajax.uninstall();
        request = undefined;
        component && component.teardown();
        component = undefined;
      });

      it('should fire successEvent on success', function () {
        request.response(testResponses.success);
        expect(component.testSuccessMethod).toHaveBeenCalled();
      });

      it('should fire errorEvent on failure', function () {
        request.response(testResponses.error);
        expect(component.testErrorMethod).toHaveBeenCalled();
      });
    });
  });
});


karma.conf.js
// Karma configuration file
//
// For all available config options and default values, see:
// https://github.com/karma-runner/karma/blob/stable/lib/config.js#L54

module.exports = function (config) {
  'use strict';

  config.set({
    // base path, that will be used to resolve files and exclude
    basePath: '',

    frameworks: [
      'jasmine',
      'requirejs'
    ],

    // list of files / patterns to load in the browser
    files: [
      // loaded without require
      'bower_components/jquery/dist/jquery.js',
      'bower_components/jasmine-jquery/lib/jasmine-jquery.js',
      'bower_components/jasmine-flight/lib/jasmine-flight.js',
      'bower_components/jasmine-ajax/lib/mock-ajax.js',

      // loaded with require
      {pattern: 'bower_components/flight/**/*.js', included: false},
      {pattern: 'lib/**/*.js', included: false},
      {pattern: 'test/spec/**/*.spec.js', included: false},

      'test/test-main.js'
    ],

    // use dots reporter, as travis terminal does not support escaping sequences
    // possible values: 'dots', 'progress'
    // CLI --reporters progress
    reporters: ['progress'],

    // enable / disable watching file and executing tests whenever any file changes
    // CLI --auto-watch --no-auto-watch
    autoWatch: true,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    // CLI --browsers Chrome, Firefox, Safari
    browsers: [
      'Chrome'
    ],

    // If browser does not capture in given timeout [ms], kill it
    // CLI --capture-timeout 5000
    captureTimeout: 20000,

    // Auto run tests on start (when browsers are captured) and exit
    // CLI --single-run --no-single-run
    singleRun: false,

    plugins: [
      'karma-jasmine',
      'karma-requirejs',
      'karma-chrome-launcher',
      'karma-firefox-launcher'
    ]
  });
};

lib/with_request.js
define(function (require) {

  'use strict';

  /**
   * Module exports
   */

  return withRequest;

  /**
   * Module function
   */

  function withRequest() {
    var requests = [];

    this.attributes({
      requestDefaultErrorMessage: 'Unknown Error',
      tunnelMethod: '_method'
    });

    this.request = function (settings, method) {
      var type = (method || settings.type || 'get').toUpperCase(),
      data = settings.data || {},
      dataType = settings.dataType || 'JSON',
      tunnelMethod = this.attr.tunnelMethod,
      request;

      if (!settings.hasOwnProperty('context')) {
        settings.context = this;
      }

      /*
        "HTML forms (up to HTML version 4 and XHTML 1) only support GET and POST as HTTP request methods. 
        A workaround for this is to tunnel other methods through POST by using a hidden form field 
        which is read by the server and the request dispatched accordingly."

        http://stackoverflow.com/a/166501/931594
      */
      if (settings.tunnelMethod !== false && type !== 'GET' && type !== 'POST') {
        if (typeof settings.tunnelMethod == 'string') {
          tunnelMethod = settings.tunnelMethod
        }

        settings[tunnelMethod] = type;

        type = 'POST';
      }

      if (settings.success) {
        settings.success = this.__requestSuccess__(settings.success, settings.context);
      }

      if (settings.error) {
        settings.error = this.__requestError__(settings.error, settings.context);
      }

      request = $.ajax($.extend(settings, {
        url: settings.url,
        type: type,
        dataType: dataType,
        data: data
      }));

      if (request && !settings.noAbort) {
        requests.push(request);
      }

      return request; 
    }

    this.__requestSuccess__ = function (action, context) {
      return function (data, textStatus, jqXHR) {
        this.__requestComplete__(jqXHR, data, action, context);
      }.bind(this);
    };

    this.__requestError__ = function (action, context) {      
      return function (jqXHR, textStatus) {
        var data = {};
        
        try {
          if (jqXHR.status !== 0) {
            data = JSON.parse(jqXHR.responseText);
          }
        } catch (e) { }

        data.status = data.status || jqXHR.status || 0;
        data.message = data.statusText || jqXHR.statusText || textStatus;


        // when a ajax request is aborted it is treated as an error
        if (data.status == 0 && jqXHR.statusText === 'abort') {
          data.isAbort = true;
        }

        if (!data.message) {
          data.message = this.attr.requestDefaultErrorMessage;
        }

        this.__requestComplete__(jqXHR, data, action, context);
      }.bind(this);
    };

    this.__requestComplete__ = function (request, responseData, action, context) {
      removeRequest(request);

      if (typeof action == 'function') {
        action.call(context, responseData);
        return;
      }

      this.trigger(action, responseData);
    };

    function removeRequest(request) {
      requests.splice(requests.indexOf(request), 1);
    }

    // mixin methods

    this.get = function (settings) {
      return this.request(settings, "GET");
    };

    this.post = function (settings) {
      return this.request(settings, "POST");
    };

    this.put = function (settings) {
      return this.request(settings, "PUT");
    };

    this.destroy = this['delete'] = function (settings) {
      return this.request(settings, "DELETE");
    };

    this.abort = function (request) {
      if (request && typeof request.abort == 'function') {
        removeRequest(request);
        request.abort();
      }
    };

    this.abortAllRequests = function () {
      requests.forEach(this.abort, null);
    };

    this.after('initialize', function () {
      this.on(window, 'unload', this.abortAllRequests);
    });
  }  
});